### PR TITLE
make baseFee even number

### PR DESCRIPTION
**Pull request by kjeom**
## Proposed changes

- Issue
  - txFee1 / 2) + (txFee2 / 2) + … + (txFeeN / 2) **!=** (txFee1 + txFee2 + … + txFeeN) / 2.
  - It is because big.Int division throws away the remainder.
  - the difference is (Count_EvenNumberTxFee / 2).
  - txFee1 / 2) + (txFee2 / 2) + … + (txFeeN / 2) **!=** (txFee1 + txFee2 + … + txFeeN) / 2 - (Count_EvenNumberTxFee / 2).
- Propose
  - make the baseFee even number by adding 1 to the OddNumberBaseFee.
  - It can be applied to the other divider (3, 4, 5) by adding the offset (divider - remainder).

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [x] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: consensus/misc/kip71.go [hunk 1]
```diff
@@ -23,14 +23,22 @@ func VerifyMagmaHeader(config *params.ChainConfig, parentHeader, header *types.H
 	return nil
 }
 
+// If the upperBoundBaseFee is Odd number then the baseFee can be upperBoundBaseFee + 1
```
[the remark this answers is not in the document]
Comment (kjeom) [reply consensus/misc/kip71.go:26]: Hmm.. I don't know what is more reasonable.
Allowing baseFee to be upperBoundBaseFee + 1 is not bad as well...

[the remark this answers is not in the document]
Comment (jimni1222) [reply consensus/misc/kip71.go:26]: It felt unnatural to me that there is an upper limit but still the base rate could exceed that value.
This was light suggestion, so we can discuss this together.

[the remark this answers is not in the document]
Comment (kjeom) [reply consensus/misc/kip71.go:26]: @bobpkr suggest the code `a.Rsh(a, 1); a.Lsh(a, 1)` that is more faster and means -1.
In this case, the baseFee can be lower than lower limit.
How about adding condition checking that the baseFee is Odd and lowerBound and statement baseFee+1?

[the remark this answers is not in the document]
Comment (jimni1222) [reply consensus/misc/kip71.go:26]: Sounds good to me!

File: consensus/misc/kip71.go [hunk 2]
```diff
@@ -23,10 +23,20 @@ func VerifyMagmaHeader(config *params.ChainConfig, parentHeader, header *types.H
 	return nil
 }
 
-// If the upperBoundBaseFee is Odd number then the baseFee can be upperBoundBaseFee + 1
-func makeEven(baseFee *big.Int) *big.Int {
+// If the upperBoundBaseFee is Odd number then the baseFee cna be upperBoundBaseFee + 1
```
Comment (jimni1222) [inline consensus/misc/kip71.go:26]: Can you fix the comment?

File: consensus/misc/kip71.go [hunk 3]
```diff
@@ -23,10 +23,29 @@ func VerifyMagmaHeader(config *params.ChainConfig, parentHeader, header *types.H
 	return nil
 }
 
+func makeEvenByDown(baseFee *big.Int) *big.Int {
+	if baseFee.Bit(0) != 0 {
+		baseFee.Rsh(baseFee, 1)
+		baseFee.Lsh(baseFee, 1)
+		return baseFee
```
Comment (bobpkr) [inline consensus/misc/kip71.go:30]:
```suggestion
```

File: consensus/misc/kip71.go [hunk 4]
```diff
@@ -23,10 +23,29 @@ func VerifyMagmaHeader(config *params.ChainConfig, parentHeader, header *types.H
...
+	}
+	return baseFee
+}
+
+func makeEvenByUp(baseFee *big.Int) *big.Int {
+	if baseFee.Bit(0) != 0 {
+		baseFee.Add(baseFee, common.Big1)
+		return baseFee
```
Comment (bobpkr) [inline consensus/misc/kip71.go:38]:
```suggestion
```

File: consensus/misc/kip71.go (hunk 3 above)
Comment (kjeom) [reply consensus/misc/kip71.go:30]: It needs return statement to use something like `return makeEvenByUp(baseFee)` in call sites.

Comment (kjeom) [reply consensus/misc/kip71.go]: ahh, ok thanks!

**Review by bobpkr — APPROVED** [review]

File: consensus/misc/kip71_test.go [hunk 5]
```diff
@@ -20,6 +20,80 @@ func getTestConfig(forkedBlockNum *big.Int) *params.ChainConfig {
...
+		{750000000000, 25000000000, 30000000, 20, 60000000, 17274858, 18432574843},
+		{750000000000, 25000000000, 30000000, 20, 60000000, 39093494, 43009212843},
+		{750000000000, 25000000000, 30000000, 20, 60000000, 18981884, 28574384343},
+		{750000000000, 25000000000, 30000000, 20, 60000000, 12873, 43009212843},
+		{750000000000, 25000000000, 30000000, 20, 60000000, 18273, 43009212843},
+		{750000000000, 25000000000, 30000000, 20, 60000000, 1, 43009212843},
+		{750000000000, 25000000000, 30000000, 20, 60000000, 18949, 43009212843},
+		{750000000000, 25000000000, 30000000, 20, 60000000, 19828, 43009212843},
+		// monkey test 1
+		{750000000001, 25123987327, 30000000, 20, 38573282, 83212345, 750000000000},
```
Comment (jimni1222) [inline consensus/misc/kip71_test.go:55]: Can we add one more case that tests with odd number lower bound base fee and parents gas used 0 (next base fee is lower bound base fee + 1)

Comment (kjeom) [reply consensus/misc/kip71_test.go]: Sure, I added {750000000000, 1, 30000000, 1, 60000000, 0, 10} and logic checking lower/upper bound

Comment (jimni1222) [reply consensus/misc/kip71_test.go]: Thank you for adding that! But can you check test code comments? The other cases are commented.

**Review by jimni1222 — APPROVED** [review]
LGTM Thanks!

**Review by bobpkr — APPROVED** [review]

File: consensus/misc/kip71.go [hunk 6]
```diff
@@ -93,6 +110,6 @@ func NextBlockBaseFee(parentHeader *types.Header, config *params.ChainConfig) *b
 		if nextBaseFee.Cmp(lowerBoundBaseFee) < 0 {
 			return lowerBoundBaseFee
 		}
-		return nextBaseFee
+		return makeEvenByDown(nextBaseFee)
```
[the remark this answers is not in the document]
Comment (aidan-kwon) [reply consensus/misc/kip71.go:113]: oh, you are right. Thanks!

**Review by aidan-kwon — APPROVED** [review]
It looks good, but we need to change the calculation logic in KIP71

File: consensus/misc/kip71.go [hunk 7]
```diff
@@ -23,10 +23,27 @@ func VerifyMagmaHeader(config *params.ChainConfig, parentHeader, header *types.H
 	return nil
 }
 
+func makeEvenByDown(baseFee *big.Int) *big.Int {
+	if baseFee.Bit(0) != 0 {
+		baseFee.Rsh(baseFee, 1)
```
Comment (kjhman21) [inline consensus/misc/kip71.go:28]: I think it would be better to use baseFee.sub()

**Comment by kjhman21** [comment]
https://github.com/klaytn/klaytn/pull/1543

Refined this code please take a look @kjeom @jimni1222 @bobpkr @aidan-kwon

**Pull request merged**